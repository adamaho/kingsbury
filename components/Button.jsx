import * as React from 'react';
import styled from '../styles/theme';
const ButtonComponent = (props) => {
    return (<button className={`kb-button ${props.className}`} {...props}>
      {props.children}
    </button>);
};
export const Button = styled(ButtonComponent) `
  background: ${(props) => props.theme.buttons[props.type].background};
  color: ${(props) => props.theme.buttons[props.type].color};
  height: ${(props) => `${props.theme.buttons.default.height}px`};

  font-family: ${(props) => props.theme.fontFamily};
  font-weight: 700;
  font-size: 16px;
  letter-spacing: 2px;
  text-decoration: none;

  border: ${(props) => props.theme.buttons[props.type].border};
  border-radius: 4px;

  padding: 5px 0px 0px 0px;
  box-shadow: ${(props) => props.theme.buttons[props.type].boxShadow};

  cursor: pointer;

  transition: border-radius 0.25s;

  &:hover {
    border-radius: ${(props) => `${props.theme.buttons.default.height / 2}px`};
  }

  &:active {
    opacity: 0.7;
  }

  &:focus {
    outline: none;
  }
`;
