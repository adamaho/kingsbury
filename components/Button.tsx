import * as React from 'react';
import * as classNames from 'classnames';
import styled from '../styles/theme';

import {
  ITheme
} from '../styles/theme';

interface IButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  type: 'disabled' | 'primary' | 'ghost';
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  className?: string;
  theme: ITheme;
}

const ButtonComponent: React.FunctionComponent<IButtonProps> = (props) => {
  return (
    <button
      className={classNames('kb-button', props.className)}
      {...props}
    >
      {props.children}
    </button>
  );
}

export const Button = styled(ButtonComponent)`
  background: ${(props: IButtonProps) => props.theme.buttons[props.type].background};
  color: ${(props: IButtonProps) => props.theme.buttons[props.type].color};
  height: ${(props: IButtonProps) => `${props.theme.buttons.default.height}px`};

  font-family: ${(props: IButtonProps) => props.theme.fontFamily};
  font-weight: 700;
  font-size: 16px;
  letter-spacing: 2px;
  text-decoration: none;

  border: ${(props: IButtonProps) => props.theme.buttons[props.type].border};
  border-radius: 4px;

  padding: 5px 0px 0px 0px;
  box-shadow: ${(props: IButtonProps) => props.theme.buttons[props.type].boxShadow};

  cursor: pointer;

  transition: border-radius 0.25s;

  &:hover {
    border-radius: ${(props: IButtonProps) => `${props.theme.buttons.default.height / 2}px`};
  }

  &:active {
    opacity: 0.7;
  }

  &:focus {
    outline: none;
  }
`
