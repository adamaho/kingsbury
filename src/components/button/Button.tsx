import * as React from 'react';

import styled, {
  css
} from 'styled-components';

import {
  MouseEventHandler
} from "react";

import {
  ThemeContext
} from "../../styled";

export interface ButtonProps {
  /** Type of the button */
  buttonType?: 'primary' | 'success' | 'danger' | 'warning' | 'info';

  /** Content to show in the button */
  children?: React.ReactNode;

  /** classname for the button */
  className?: string;

  /** Disabled state of the button */
  disabled?: boolean;

  /** Function to handle click event */
  onClick?: MouseEventHandler<HTMLButtonElement>;

  /** Ref to be passed to the button */
  ref?: React.Ref<HTMLButtonElement>;
}

const defaultButtonType = 'primary';

const REQUIRE_DARK_TEXT = new Set(['success', 'warning']);

const StyledButton = styled.button<ButtonProps>`
  background: ${(props) => (
    props.disabled ?
      props.theme.colors.disabled :
      props.theme.colors[props.buttonType || defaultButtonType]
  )};

  color: ${(props) => {
    if (REQUIRE_DARK_TEXT.has(props.buttonType || defaultButtonType)) {
      return props.theme.colors.black;
    }
    return props.theme.colors.white;
  }};

  height: ${(props) => `${props.theme.button.height}`};

  font-weight: 400;
  font-size: 16px;
  font-family: inherit;
  text-decoration: none;

  border: ${(props) => props.theme.button.border};
  border-radius: ${(props) => props.theme.button.borderRadius};

  padding: ${(props) => props.theme.button.padding};

  cursor: ${(props) => (
    props.disabled ? 'not-allowed' : 'pointer'
  )};

  transition: ${(props) => props.theme.animations.time.fast} cubic-bezier(.19, 1, .4, 1);

  ${(props) => !props.disabled && css`
    &:hover {
      opacity: 0.85;
    }
  `}

  &:active {
    opacity: 1;
  }

  &:focus {
    outline: none;
  }
`;

export const Button: React.FunctionComponent<ButtonProps> = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const {
    children,
    buttonType
  } = props;

  const theme = React.useContext(ThemeContext);

  return (
    <StyledButton
      ref={ref}
      buttonType={buttonType}
      theme={theme}
      {...props}
    >
      {children}
    </StyledButton>
  );
});

Button.defaultProps = {
  buttonType: 'primary',
  children: '',
  className: '',
  disabled: false,
  onClick: undefined
};
