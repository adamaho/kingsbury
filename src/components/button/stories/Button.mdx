import { Story, Preview, Props, Meta } from '@storybook/addon-docs/blocks';
import {Button} from '../Button';

# Button

I can embed a story (but not define one, since this file should not contain a `Meta`):

<Preview>
  <Story id="components-button--basic" />
</Preview>

<Props of={Button} />
