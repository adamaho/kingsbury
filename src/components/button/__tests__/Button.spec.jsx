import React from 'react';
import enzyme from 'enzyme';
import renderer from 'react-test-renderer';

import Button from '..';

describe('Button', () => {
  test('it renders without children', () => {
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders with children', () => {
    const tree = renderer.create(
      <Button>
        test
      </Button>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it calls onClick', () => {
    const onClickMock = jest.fn();
    const component = enzyme.mount(
      <Button onClick={onClickMock} />
    );

    component.find('button').simulate('click');
    expect(onClickMock).toHaveBeenCalled();
  });
});
