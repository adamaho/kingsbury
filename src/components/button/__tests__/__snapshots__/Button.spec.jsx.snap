// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button it calls onClick 1`] = `
.c0 {
  background: #5C6AC4;
  color: #FFFFFF;
  height: 38px;
  font-weight: 400;
  font-size: 16px;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
  border-radius: 4px;
  padding: 5px 15px 5px 15px;
  cursor: pointer;
  -webkit-transition: 0.25s cubic-bezier(.19,1,.4,1);
  transition: 0.25s cubic-bezier(.19,1,.4,1);
}

.c0:hover {
  opacity: 0.85;
}

.c0:active {
  opacity: 1;
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
  disabled={false}
  onClick={[Function]}
>
  test
</button>
`;

exports[`Button it calls onClick 2`] = `
<button
  className="button__StyledButton-NiLSb jeuqvP"
  disabled={false}
  onClick={[Function]}
>
  test
</button>
`;

exports[`Button it renders with children 1`] = `
.c0 {
  background: #5C6AC4;
  color: #FFFFFF;
  height: 38px;
  font-weight: 400;
  font-size: 16px;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
  border-radius: 4px;
  padding: 5px 15px 5px 15px;
  cursor: pointer;
  -webkit-transition: 0.25s cubic-bezier(.19,1,.4,1);
  transition: 0.25s cubic-bezier(.19,1,.4,1);
}

.c0:hover {
  opacity: 0.85;
}

.c0:active {
  opacity: 1;
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
  disabled={false}
>
  test
</button>
`;

exports[`Button it renders without children 1`] = `
.c0 {
  background: #5C6AC4;
  color: #FFFFFF;
  height: 38px;
  font-weight: 400;
  font-size: 16px;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
  border-radius: 4px;
  padding: 5px 15px 5px 15px;
  cursor: pointer;
  -webkit-transition: 0.25s cubic-bezier(.19,1,.4,1);
  transition: 0.25s cubic-bezier(.19,1,.4,1);
}

.c0:hover {
  opacity: 0.85;
}

.c0:active {
  opacity: 1;
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
  disabled={false}
>
  
</button>
`;
